packages/@ourworldindata/utils/src/profiles.ts

```diff
@@ -0,0 +1,179 @@
+    replacements: ReplacementPair[]
+): void => {
+    for (const key of Object.keys(record)) {
+        const value = record[key]
+        if (typeof value === "string") {
+            record[key] = replaceTokens(value, replacements)
+        } else if (Array.isArray(value)) {
+            replaceStringsInArray(value, replacements)
+        } else if (value && typeof value === "object") {
+            replaceStringsInObject(value as Record<string, any>, replacements)
+        }
+    }
+}
+
+/**
+ * Instantiates a profile content by replacing tokens with entity-specific values.
+ */
+export const instantiateProfile = (
+    content: OwidGdocProfileContent,
+    entity: ProfileEntity
+): OwidGdocProfileContent => {
+    const entityNameWithArticle = articulateEntity(entity.name)
+    const replacements: ReplacementPair[] = [
+        ["$entityName", entityNameWithArticle],
+        ["$entityCode", entity.code],
+        ["$noArticleEntityName", entity.name],
+    ]
+
+    const clonedContent = _.cloneDeep(content)
+
+    clonedContent.title = replaceTokens(clonedContent.title, replacements)
+    clonedContent.subtitle = applyToOptionalString(
+        clonedContent.subtitle,
+        replacements
+    )
+    if (clonedContent.excerpt) {
+        clonedContent.excerpt = replaceTokens(
+            clonedContent.excerpt,
+            replacements
+        )
+    }
+    if (clonedContent.toc) {
+        replaceStringsInObject(clonedContent.toc, replacements)
+    }
+    if (clonedContent.refs) {
+        replaceStringsInObject(clonedContent.refs, replacements)
+    }
+    if (clonedContent.body) {
+        replaceStringsInObject(clonedContent.body, replacements)
+    }
+
+    const region = getRegionByNameOrVariantName(entity.name)
+    if (region) {
+        clonedContent.instantiatedEntity = {
+            ...region,
+            isCountry: checkIsCountry(region),
+        }
+    }
+
+    return clonedContent
+}
+
+export function getEntitiesForProfile(
+    profileTemplate: OwidGdocProfileInterface
+): ProfileEntity[] {
+    const scopeRaw = profileTemplate.content.scope
+    if (!scopeRaw) return []
+
+    const scopeValues = scopeRaw
+        .split(",")
+        .map((value) => value.trim().toLowerCase())
+        .filter((value) => value.length > 0)
+
+    const entitiesByCode = new Map<string, ProfileEntity>()
+    const addEntity = (region?: { name?: string; code?: string }): void => {
+        if (!region?.code || !region?.name) return
+        if (!entitiesByCode.has(region.code)) {
+            entitiesByCode.set(region.code, {
+                name: region.name,
+                code: region.code,
+            })
+        }
+    }
+
+    const nonCountryRegions = regions.filter(
+        (region) => !checkIsCountry(region)
+    )
+
+    for (const scopeValue of scopeValues) {
+        if (scopeValue === "countries" || scopeValue === "all") {
+            for (const country of countries) addEntity(country)
```

Comment:
We also have a special "Other" category that contains unrecognised countries like Somaliland. Not saying it should be included, just wanted to make you aware!